consensus/istanbul/types.go

```diff
@@ -343,6 +425,65 @@ type Message struct {
 	Msg       []byte
 	Address   common.Address // The sender address
 	Signature []byte         // Signature of the Message using the private key associated with the "Address" field
+
+	// The below fields are not serializable since they are private, they are
+	// set when calling Message.FromPayload, only one will be set in any
+	// instance, which is set depends on the Message.Code.
+	committedSubject    *CommittedSubject
+	prePrepare          *Preprepare
+	prepare             *Subject
+	roundChange         *RoundChange
+	queryEnode          *QueryEnodeData
+	forwardMessage      *ForwardMessage
+	enodeCertificate    *EnodeCertificate
+	versionCertificates []*VersionCertificate
+	valEnodeShareData   *ValEnodesShareData
```

Comment:
This seems to be a prime use-case for interfaces. Is there a reason to prefer this approach over piecing this out into an interface and various implementations?

Reply:
Could you give an example, I'm not seeing how to use an interface here?

Reply:
Maybe you could do
```
type Message interface {
       Msg()       []byte
      Address()  common.Address // The sender address
      Signature() []byte         // Signature of the Message using the private key associated with the "Address" 
}
```
And then have a struct for message that fulfills the interface.

One general comment I have by looking at this code is that while I don't know how easy it is, I'd like to be able to have have a `parseMessage` which returns something I can do a typeswitch on to handle each message (so each handler is guaranteed by the type system to have the correct type). Unfortunately go doesn't make this particularly easy.

Reply:
@trianglesphere, I'm not really seeing what that interface gets you, it seems to be just repeating what is already available on an instance of instanbul.Message minus the `Code` field.
```go
type Message struct {
	Code      uint64
	Msg       []byte
	Address   common.Address // The sender address
	Signature []byte         // Signature of the Message using the private key associated with the "Address" field
}

```

Reply:
What find uncomfortable about this factoring is that there are a number of private fields, one for each type of message, and the reliance on mutations. It also doesn't provide a clear way to make sure a message is a particular type, which can result in bugs, like the `nil` pointer error I mentioned in another comment.